src/lib.rs

```diff
@@ -1264,6 +1277,80 @@ where
     }
 }
 
+/// A streaming iterator that flattens nested streaming iterators.
+#[derive(Debug)]
+pub struct Flatten<I> {
+    iter: I,
+    first: bool,
+}
+
+impl<I, J> StreamingIterator for Flatten<I>
+where
+    I: StreamingIteratorMut<Item = J>,
+    for<'a> J: StreamingIterator + 'a,
+{
+    type Item = J::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        if self.first {
+            self.first = false;
+            self.iter.advance();
+        }
+        while let Some(iter) = self.iter.get_mut() {
+            iter.advance();
+            if !iter.is_done() {
+                break;
+            }
+            self.iter.advance(); // since we got Some, self.iter is not done and can be advanced
```

Comment:
The pre-NLL borrowck didn't like this, mutating `self.iter` while `get_mut()` was lexically alive. I moved the crate to 2018 edition in #29, so now we do get NLL here and it's happy.